src/System.Management.Automation/engine/Subsystem/Prediction/CommandPrediction.cs

```diff
@@ -0,0 +1,174 @@
+using System.Threading.Tasks;
+
+namespace System.Management.Automation.Subsystem
+{
+    /// <summary>
+    /// The class represents the prediction result from a predictor.
+    /// </summary>
+    public sealed class PredictionResult
+    {
+        /// <summary>
+        /// Gets the Id of the predictor.
+        /// </summary>
+        public Guid Id { get; }
+
+        /// <summary>
+        /// Gets the suggestions.
+        /// </summary>
+        public List<string> Suggestions { get; }
+
+        internal PredictionResult(Guid id, List<string> suggestions)
+        {
+            Id = id;
+            Suggestions = suggestions;
+        }
+    }
+
+    /// <summary>
+    /// Provides a set of possible predictions for given input.
+    /// </summary>
+    public static class CommandPrediction
+    {
+        /// <summary>
+        /// Collect the predictive suggestions from registered predictors using the default timeout.
+        /// </summary>
+        /// <param name="ast">The <see cref="Ast"/> object from parsing the current command line input.</param>
+        /// <param name="astTokens">The <see cref="Token"/> objects from parsing the current command line input.</param>
+        /// <returns>A list of <see cref="PredictionResult"/> objects.</returns>
+        public static Task<List<PredictionResult>> PredictInput(Ast ast, Token[] astTokens)
+        {
+            return PredictInput(ast, astTokens, millisecondsTimeout: 20);
+        }
+
+        /// <summary>
+        /// Collect the predictive suggestions from registered predictors using the specified timeout.
+        /// </summary>
+        /// <param name="ast">The <see cref="Ast"/> object from parsing the current command line input.</param>
+        /// <param name="astTokens">The <see cref="Token"/> objects from parsing the current command line input.</param>
+        /// <param name="millisecondsTimeout">The milliseconds to timeout.</param>
+        /// <returns>A list of <see cref="PredictionResult"/> objects.</returns>
+        public static async Task<List<PredictionResult>> PredictInput(Ast ast, Token[] astTokens, int millisecondsTimeout)
+        {
+            var predictors = SubsystemManager.GetSubsystems<IPredictor>();
+            if (predictors.Count == 0)
+            {
+                return null;
+            }
+
+            var context = new PredictionContext(ast, astTokens);
+            var cancellationSource = new CancellationTokenSource();
+            var cancellationToken = cancellationSource.Token;
+            var tasks = new Task<PredictionResult>[predictors.Count];
+
+            for (int i = 0; i < predictors.Count; i++)
+            {
+                var predictor = predictors[i];
+
+                tasks[i] = Task.Factory.StartNew(
+                    state => {
+                        var predictor = (IPredictor)state;
+                        List<string> texts = predictor.GetSuggestion(context, cancellationToken);
```

Comment:
Since the implementer can't be PowerShell anyway, should this API also be `async`?  If it's synchronous and the external code ends up calling `async` API's then you might risk thread starvation in environments with low thread pool sizes.

Reply:
I thought about making it `async`, but didn't go that way.
Here is my reasoning: there is a strict time limit on how long this method should run to finish -- 20 ms by default for now, so doing async operations in this method feels like wasting time on extra overheads (scheduling, context switching and etc.). That's why I chose to make it synchronous, kinda like a suggestion/hint that synchronous operations are preferred in this method implementation.